### ISSUE TITLE
qq.s3.XhrUploadHandler._getLocalStorageId is never called 

### ISSUE THREAD

**Issue by e-tip** [FineUploader/fine-uploader#1617]
## Type of issue [REQUIRED]
bug

## Uploader type [REQUIRED]
S3

### Fine Uploader version [REQUIRED]
5.10.1 and previous

### Browsers where the bug is reproducible [REQUIRED]
Chrome , Firefox and Safari

### Operating systems where the bug is reproducible [REQUIRED]
OSX 10.11

### Detailed explanation of the problem [REQUIRED]
I wanto to generate an localStorageId that keeps track of the full s3 key instead of the filename only, but i've noticed that _getLocalStorageId is never overridden even if in  qq.s3.XhrUploadHandler there's this code 
`_getLocalStorageId: function(id) {
                var baseStorageId = super_._getLocalStorageId(id),
                    bucketName = upload.bucket.getName(id);

                return baseStorageId + "-" + bucketName;
            }`

**Comment by rnicholus** [comment]
Please include code that reproduces whatever issue you are having.

**Comment by e-tip** [comment]
i'll setup a demo as soon as possible

**Comment by e-tip** [comment]
Here it's [http://labs.e-tip.net/](url)
it's just the demo code. the only modification i've done is 
at line 10697 of s3.fine-uploader.js 
```javascript
_getLocalStorageId: function(id) {
                var baseStorageId = super_._getLocalStorageId(id),
                    bucketName = upload.bucket.getName(id);
                    qq.log(baseStorageId + "-" + bucketName);
                return baseStorageId + "-" + bucketName;
            }
```
and at line 5526 
```javascript
_getLocalStorageId: function(id) {
            var formatVersion = "5.0",
                name = getName(id),
                size = getSize(id),
                chunkSize = chunking.partSize,
                endpoint = getEndpoint(id);
            qq.log(qq.format("qq{}resume{}-{}-{}-{}-{}", namespace, formatVersion, name, size, chunkSize, endpoint));
            return qq.format("qq{}resume{}-{}-{}-{}-{}", namespace, formatVersion, name, size, chunkSize, endpoint);
        },
```
As soon i add a file i can see in the console that localStorageId is always returned by function at line 5526 because i can see no -nucketName appended

**Comment by rnicholus** [comment]
What specific issue are you seeing? Is something broken or not functioning correctly? For example: "The resume feature is not working if x, y, and z" If so, please provide specific steps needed to reproduce a noticeable issue.

**Comment by e-tip** [comment]
Hi, i've modified my test suite.
i've added a second uploader and modified the uploader code
for first uploader i've added to objectProperties object the following key function
```javascript
key : function(fileID){
    return "folder1/"+uploader.getUuid(fileID)+"."+qq.getExtension(uploader.getName(fileID));
}
```
and for second uploader 
```javascript
key : function(fileID){
    return "folder2/"+uploader.getUuid(fileID)+"."+qq.getExtension(uploader.getName(fileID));
}
```
Steps to reproduce the issue
- Start uploading a file larger than 6 MB to uploader1 so chunk upload is enabled
- upload endpoint is this http://fineuploader-test.s3.eu-central-1.amazonaws.com/folder1/59950077-0b84-4ca6-9f5a-d242fda00f2e.mov
- After upload is started refresh the page so upload is interrupted
- Drop the same file in the second uploader and you'll see that the file is uploaded in folder1 even if key function points to folder2. 
This happens because the localStorageID is calculated not considering that the key can be different for the same filename-filesize keys ( that is the actual localStorageId calculating method )

**Comment by rnicholus** [comment]
Thanks for the explanation. Can you open up a pull request with your fix so we can discuss that further?

**Comment by e-tip** [comment]
I'll do as soon as i understand why qq.XhrUploadHandler._getLocalStorageId is called instead of qq.s3.XhrUploadHandler._getLocalStorageId

**Comment by e-tip** [comment]
I'll try to write a fix and open up a pull request. By the way the main issue is that the add function is called in qq.XhrUploadHandler contest so handler. is kind of qq.XhrUploadHandler instead of qq.s3.XhrUploadHandler. as soon as i fix this issue my solution would be to check if objectProperties.key is a function and if yes check if returns a kind of string that contains /. if yes strip out the filename and use also that part as localStorage key. Does that sound good for you ?

**Comment by rnicholus** [comment]
There is a lot if inheritance between the modules, so all may. It be as it seems. I plan to look at this closer soon.

**Comment by e-tip** [comment]
yep, actually the main issue i'm having is due to inheritance between s3 and non s3 version, i came out with too many methods to override

**Comment by rnicholus** [comment]
One approach is to simply append the key name onto the generated localStorageId in the S3-specific module. In fact, I wouldn't suggest making any changes outside of s3.xhr.upload.handler.js. The only potentially complex part here is that key determination may be an async operation. Currently, that key is first resolved when the upload begins, but that would have to change here. Also, any change to the local storage ID would potentially invalidate all existing localStorageId records, and that is not acceptable in my opinion. So code would need to be added to distinguish between the current v5.0 localStorageIds and this new version (5.?).

**Comment by e-tip** [comment]
I agree with you that invalidating all actual storageIds is not acceptable. 
i've played today with sources to try appending some custom code to the localStorageId that is generated in 3.xhr.upload.handler.js but that function is never called. I have to play some more. Thanks for checking into this for me :)
